agimus_controller/agimus_controller/ocp_base_croco.py

```diff
@@ -0,0 +1,133 @@
+
+        Args:
+            robot_model (RobotModelFactory): All models of the robot.
+            ocp_params (OCPParamsBaseCroco): Input data structure of the OCP.
+        """
+        # Setting the robot model
+        self._robot_model = robot_model
+        self._rmodel = self._robot_model._rmodel
+        self._cmodel = self._robot_model._complete_collision_model
+        self._armature = self._robot_model._params.armature
+
+        # Stat and actuation model
+        self._state = crocoddyl.StateMultibody(self._rmodel)
+        self._actuation = crocoddyl.ActuationModelFull(self._state)
+
+        # Setting the OCP parameters
+        self._ocp_params = ocp_params
+
+        self._ocp_results = None
+
+    @property
+    def horizon_size(self) -> int:
+        """Number of time steps in the horizon."""
+        return self._ocp_params.horizon_size
+
+    @property
+    def dt(self) -> float:
+        """Integration step of the OCP."""
+        return self._ocp_params.dt
+
+    @abstractmethod
+    def create_running_model_list(self) -> list[crocoddyl.ActionModelAbstract]:
+        """Create the list of running models."""
+        pass
+
+    @abstractmethod
+    def create_terminal_model(self) -> crocoddyl.ActionModelAbstract:
+        """Create the terminal model."""
+        pass
+
+    def solve(
+        self,
+        x0: npt.NDArray[np.float64],
+        x_warmstart: list[npt.NDArray[np.float64]],
+        u_warmstart: list[npt.NDArray[np.float64]],
+    ) -> None:
+        """Solves the OCP.
+        The results can be accessed through the ocp_results property.
+
+        Args:
+            x0 (npt.NDArray[np.float64]): Current state of the robot.
+            x_warmstart (list[npt.NDArray[np.float64]]): Predicted states for the OCP.
+            u_warmstart (list[npt.NDArray[np.float64]]): Predicted control inputs for the OCP.
+        """
+        ### Creation of the state and actuation models
+        # Create the running models
+        self.running_model_list = self.create_running_model_list()
+        # Create the terminal model
+        self.terminal_model = self.create_terminal_model()
+
+        # Create the shooting problem
+        problem = crocoddyl.ShootingProblem(
+            x0, self.running_model_list, self.terminal_model
+        )
+        # Create solver + callbacks
+        ocp = mim_solvers.SolverCSQP(problem)
+
+        # Merit function
+        ocp.use_filter_line_search = self._ocp_params.use_filter_line_search
+
+        # Parameters of the solver
+        ocp.termination_tolerance = self._ocp_params.termination_tolerance
+        ocp.max_qp_iters = self._ocp_params.qp_iters
+        ocp.eps_abs = self._ocp_params.eps_abs
+        ocp.eps_rel = self._ocp_params.eps_rel
+        ocp.with_callbacks = self._ocp_params.callbacks
+
+        # Creating the warmstart lists for the solver
+        # Solve the OCP
+        ocp.solve([x0] + x_warmstart, u_warmstart, self._ocp_params.solver_iters)
```

Comment:
This way we instantiate the problem and the solver only once, btw isn't x0 already in x_warmstart ?
```suggestion
        if self._ocp is None:
            self._problem = crocoddyl.ShootingProblem(
                x0, self.running_model_list, self.terminal_model
            )
            # Create solver + callbacks
            self._ocp = mim_solvers.SolverCSQP(self._problem)
    
            # Merit function
            self._ocp.use_filter_line_search = self._ocp_params.use_filter_line_search
    
            # Parameters of the solver
            self._ocp.termination_tolerance = self._ocp_params.termination_tolerance
            self._ocp.max_qp_iters = self._ocp_params.qp_iters
            self._ocp.eps_abs = self._ocp_params.eps_abs
            self._ocp.eps_rel = self._ocp_params.eps_rel
            self._ocp.with_callbacks = self._ocp_params.callbacks

        # Creating the warmstart lists for the solver
        # Solve the OCP
        self._ocp.solve([x0] + x_warmstart, u_warmstart, self._ocp_params.solver_iters)
```

Reply:
Good point, thank you!

Reply:
x0 and x_warmstart[0] may be different depending on how x_warmstart was computed, I would keep them separated. Example: x0 read from the joint sensors vs x_warmstart obtained by shifting the previous solution.